src/tools/sc-hsm-tool.c

```diff
@@ -1667,6 +1693,298 @@ static int unwrap_key(sc_card_t *card, int keyid, const char *inf, const char *p
+		return -1;
+	}
+
+	if (fwrite(out, 1, outlen, outfp) != (size_t)outlen) {
+		perror(outf);
+		fclose(outfp);
+		free(out);
+		free(data);
+		return -1;
+	}
+
+	fclose(outfp);
+	free(out);
+	free(data);
+	return 0;
+}
+
+static void get_CHR(char *chrstr, int is_cvc, sc_context_t *ctx, const u8 *buf, size_t buflen)
+{
+	const u8 *cv = NULL, *cb = NULL, *chr = NULL;
+	size_t taglen;
+
+	/* find CV certificate (should be right at the start) */
+	if (is_cvc) {
+		if (!(cv = sc_asn1_find_tag(ctx, buf, buflen, 0x7F21, &taglen))) {
+			strcpy(chrstr, "(CV certificate not found)");
+			return;
+		}
+		buf = cv;
+		buflen = taglen;
+	}
+
+	/* find embedded Certificate Body */
+	if (!(cb = sc_asn1_find_tag(ctx, buf, buflen, 0x7F4E, &taglen))) {
+		strcpy(chrstr, "(Certificate Body not found)");
+		return;
+	}
+	buf = cb;
+	buflen = taglen;
+
+    /* find embedded Certification Holder Reference (CHR) */
+	if (!(chr = sc_asn1_find_tag(ctx, buf, buflen, 0x5F20, &taglen))) {
+		strcpy(chrstr, "(CHR not found)");
+		return;
+	}
+
+	/* return CHR */
+	strncpy(chrstr, (const char*) chr, taglen);
+}
+
+
+static int register_public_key_with_card(sc_context_t *ctx, sc_card_t *card, const u8 *pk, size_t pk_len, const u8 *devcert, size_t devcert_len, const u8 *dicacert, size_t dicacert_len)
+{
+	sc_cardctl_sc_hsm_public_key_t public_key;
+	char pk_chr[BUFSIZ] = { 0 }, devcert_chr[BUFSIZ] = { 0 }, dicacert_chr[BUFSIZ] = { 0 };
+	int r;
+
+	get_CHR(pk_chr, 1, ctx, pk, pk_len);
+	get_CHR(devcert_chr, 0, ctx, devcert, devcert_len);
+	fprintf(stderr, "Adding the key issued to '%s' on device '%s'.\n", pk_chr, devcert_chr);
+	get_CHR(dicacert_chr, 0, ctx, dicacert, dicacert_len);
+
+	public_key.pk = pk;
+	public_key.pk_length = pk_len;
+	public_key.devcert = devcert;
+	public_key.devcert_length = devcert_len;
+	public_key.dicacert = dicacert;
+	public_key.dicacert_length = dicacert_len;
+	public_key.devcert_chr = (const u8*) devcert_chr;
+	public_key.devcert_chr_length = strlen(devcert_chr);
+	public_key.dicacert_chr = (const u8*) dicacert_chr;
+	public_key.dicacert_chr_length = strlen(dicacert_chr);
+
+	r = sc_card_ctl(card, SC_CARDCTL_SC_HSM_REGISTER_PUBLIC_KEY, (void *)&public_key);
+	if (r == SC_ERROR_INS_NOT_SUPPORTED) { /* Not supported or not initialized for public key registration */
+		fprintf(stderr, "Card not initialized for public key registration\n");
+		return -1;
+	}
+	if (r < 0) {
+		fprintf(stderr, "sc_card_ctl(*, SC_CARDCTL_SC_HSM_REGISTER_PUBLIC_KEY, *) failed with %s\n", sc_strerror(r));
+		return -1;
+	}
+	fprintf(stderr, "Done.\n");
+	return 0;
+}
+
+
+
+static int register_public_key(sc_context_t *ctx, sc_card_t *card, const char *inf)
+{
+	FILE *in;
+	struct stat sb;
+	u8 *pka;
+	u8 tag = SC_ASN1_TAG_CONSTRUCTED | SC_ASN1_TAG_SEQUENCE; /* 0x30 */
+	unsigned int cla_out, tag_out;
+	const u8 *buf;
+	const u8 *pk;
+	const u8 *devcert;
+	const u8 *dicacert;
+	size_t taglen, pk_len, devcert_len, dicacert_len;
+	int r;
+
+	if (!(in = fopen(inf, "rb"))) {
```

Comment:
you may want to use `fread_to_eof()`

Reply:
Done, but it required build system changes. Would be great if you could review them.

Reply:
I think I rather stick with the current code, I still need the size of the read buffer and using `fread_to_eof()` would also require build system changes.

Reply:
```
int fread_to_eof(const char *file, unsigned char **buf, size_t *buflen);
```
`buflen` will contain the length of the allocated buffer, i.e. the read file.

The "build system changes" are as simple as adding *fread_to_eof.c* to sc-hsm-tool.

On the other hand, using `fread_to_eof()` would have saved me reviewing 50 lines of code and will save us maintaining them in the future.